docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           return new Address(
+               $value->get('street'),
+               $value->get('postCode'),
+               $value->get('city'),
+               $value->get('country'),
+           );
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               'street' => $value->street,
+               'postCode' => $value->postCode,
+               'city' => $value->city,
+               'country' => $value->country,
+           ]);
+       }
+   }
+
+This codec is quite similar to the ``PersonCodec`` we had before, except that we're not adding an identifier to it. The
+creation of the object also differs, as we have to pass all the fields to the constructor.
+
+In the ``PersonCodec``, we can now extend the ``decode`` and ``encode`` methods to handle the address field. Note that
+the example below excludes some code we've already shown in previous examples.
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       public function __construct(
+           private readonly AddressCodec $addressCodec = new AddressCodec(),
+       ) {}
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+
+           if ($value->has('address')) {
+               $person->address = $this->addressCodec->decode($value->get('address'));
+           }
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           $data = [
+               '_id' => $value->id,
+               'name' => $value->name,
+           ];
+
+           if ($value->address) {
+               $data['address'] = $this->addressCodec->encode($value->address);
+           }
+
+           return MongoDB\BSON\Document::fromPHP($data);
+       }
+   }
+
+Codec Libraries
+---------------
+
+If you have a number of codecs that you want to use in multiple places, you can create a codec library. A codec library
+contains a list of codecs and checks each codec if it supports a value. If it does, it will use that codec to encode or
+decode the given value. The following code snippet changes the ``PersonCodec`` to use a codec library instead of a
+hard-coded ``DateTimeCodec``:
```

Comment:
Should you clarify that a CodecLibrary is, itself, a codec, and accepted everywhere a singular codec can be provided?

> A codec library contains a list of codecs and presents itself as a codec. When encoding or decoding a value, it will consult each registered codec in sequence and delegate to the first supporting codec to encode or decode that value.
>
> The following code snippet changes the ``PersonCodec`` to use a codec library instead of a...

Intentionally broke up the paragraph for readability. In general, I think if you're explaining some complex topic (or introducing something) in a paragraph and then leading into an example, the example-preceding blurb is best written on its own.